### PR TITLE
Add support for HYBRID_SHARD and _HYBRID_SHARD_ZERO2 in the trainer

### PR DESCRIPTION
**Pull request by raghavanone**
#21156

### CODE REVIEW

**Comment by raghavanone** [comment]
@sgugger   Requesting for review.

File: src/transformers/trainer_utils.py [hunk 1]
```diff
@@ -663,6 +663,8 @@ class FSDPOption(ExplicitEnum):
     NO_SHARD = "no_shard"
     OFFLOAD = "offload"
     AUTO_WRAP = "auto_wrap"
+    HYBRID_SHARD = "hybrid_shard"
+    _HYBRID_SHARD_ZERO2 = "_hybrid_shard_zero2"
```
Comment (sgugger) [inline src/transformers/trainer_utils.py:667]: Let's not name public facing API with a name beginning with an _

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -422,6 +422,8 @@ class TrainingArguments:
             - `"offload"`: Offload parameters and gradients to CPUs (only compatible with `"full_shard"` and
               `"shard_grad_op"`).
             - `"auto_wrap"`: Automatically recursively wrap layers with FSDP using `default_auto_wrap_policy`.
+            - `"hybrid_shard"`: Apply `full_shard` within a node, and replicate parameters across nodes`.
+            - `"_hybrid_shard_zero2"`: Apply `shard_grad_op` within a node, and replicate parameters across nodes.
```
Comment (sgugger) [inline src/transformers/training_args.py:426]: Same here, if it's public it doesn't need an `_`.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR. Could you also have a look @pacman100 ?

**Comment by raghavanone** [comment]
@pacman100 Gentle remainder.

**Comment by pacman100** [comment]
Hello @raghavanone, after PR #23158, the FSDP logic is being handled by Accelerate. So, these changes so reflect in Accelerate if they aren't being supported by FSDP-XLA

**Comment by pacman100** [comment]
Hello @raghavanone, the above PR adds this functionality to the Accelerate which now powers Trainer—also made you the co-author of it. Thank you for all the effort and helping us keep up-to-date with the PyTorch FSDP.

**Comment by uygnef** [comment]
> Hello @raghavanone, after PR #23158, the FSDP logic is being handled by Accelerate. So, these changes should reflect in Accelerate if they aren't being supported by FSDP-XLA

Hi @pacman100, I have a question. Does this PR #23158 only work for accelerating launch? I am trying to use hybrid shared by using torchrun with trainer args "torchrun --fsdp hybrid shared ...". It seems that I need to merge #23812. Is there a better solution without modifying the transformers code?

**Pull request closed without merging**